Make terraform+CockroachDB work on GCE

cuongdo opened the pull request:
This updated terraform config allows one to create an insecure
CockroachDB cluster on GCE.

Caveats:
- Secure clusters are NOT supported.
- I didn't investigate the load balancer issues at all, so that's
  probably still broken.
- There's no supervisor config for running `cockroach`.
- I'm new to terraform, so I'm probably not doing things particularly
  elegantly.

Preview the updated `README.md` here:
https://github.com/cuongdo/cockroach/blob/terraform_gce/cloud/gce/README.md

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5890)

<!-- Reviewable:end -->

cuongdo committed:
Make terraform+CockroachDB work on GCE

This updated terraform config allows one to create an insecure
CockroachDB cluster on GCE.

Caveats:
- Secure clusters are NOT supported.
- I didn't investigate the load balancer issues at all, so that's
  probably still broken.
- There's no supervisor config for running `cockroach`.

tschottdorf commented:
Looks good (especially the sample file, and the readme updates), and seems to work so far! I'm saying "so far" because it's currently (and will be for a long time) busy uploading 100mb through my fairly slow upstream at home. While it can be more flexible to deploy a homegrown binary, I think the primary use case should be what the aws terraform stuff does: download something from S3 (we're all on OSX and upstream is going to be an issue). Maybe both ways can be supported.

I'll leave the terraform-whizzing to @mberhault, but one comment that I have is that we should put some effort into not having divergent feature sets for `AWS` and `GCE` - they should "work" kind of the same.

tschottdorf commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cloud/gce/config.sh.tpl, line 5 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEk3EXHaGtLJ3WMjNAe:-KEk3EXHaGtLJ3WMjNAf:-300239633) ([raw file](https://github.com/cockroachdb/cockroach/blob/b59c3de76819cdf2c22625952d2dbc2e705b661e/cloud/gce/config.sh.tpl#L5)):*
newline

---

*[cloud/gce/samples/3_node_cluster.tfvars, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEk3_KL1D0LSj4acNE9:-KEk3_KL1D0LSj4acNEA:-1591921998) ([raw file](https://github.com/cockroachdb/cockroach/blob/b59c3de76819cdf2c22625952d2dbc2e705b661e/cloud/gce/samples/3_node_cluster.tfvars#L34)):*
newline. Editor config perhaps?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Can the `sql_url` output variable return all of the urls? Same for `admin_url` (they also don't seem to be using the same host, at least when I just ran it).

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KEk7ulDiTVHDOepSOvv:2115385761)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
When I ran `terraform destroy` (incorrectly without args) I got this error:

```
* file: open /Users/tschottdorf/terraform/cockroach-marc-64455dfdb138.json: no such file or directory
```

which suggests that somewhere in the terraform files, something like that is hard coded and should be removed.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KEk8IYr8B9Rr_ck8rQm:1636310566)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@tschottdorf: I've removed `gce_project` and `gce_account_file` from `variables.tf`, so that `terraform destroy` prompts for those variables instead of having nonsensical default values. That should address your `terraform destroy` issue.

Regarding the URLs, they were pointing to different load balancers. I've fixed that and have also added the individual instance URLs in case there's some debugging to be done. Here's sample output:

```
  admin_lb_url = http://104.196.53.130:8080/
  admin_urls   = http://104.196.63.42:8080/, http://104.196.139.40:8080/, http://104.196.17.15:8080/, http://104.196.119.186:8080/, http://104.196.100.216:8080/
  instances    = cockroach-0, cockroach-1, cockroach-2, cockroach-3, cockroach-4
  sql_lb_url   = postgresql://root@104.196.53.130:26257/?sslmode=disable
  sql_urls     = postgresql://root@$104.196.63.42:26257/?sslmode=disable, postgresql://root@$104.196.139.40:26257/?sslmode=disable, postgresql://root@$104.196.17.15:26257/?sslmode=disable, postgresql://root@$104.196.119.186:26257/?sslmode=disable, postgresql://root@$104.196.100.216:26257/?sslmode=disable
```

I've also made the configs download the latest `cockroach` binary by default, with the option of using your own binary by overriding the `cockroach_binary` variable.

It's worth noting that bringing up clusters with terraform & GCE is a little flaky until #5916 makes it into an official build. Terraform performs operations in parallel, and the Gossip join was failing and never retrying if a node came up before the node it was trying to join. I've verified that manually re-running `launch.sh` on the affected nodes allows `cockroach` to start properly.

---

Review status: 2 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[cloud/gce/config.sh.tpl, line 5 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEk3EXHaGtLJ3WMjNAe:-KEl6_-IsW8_soXm4Och:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b59c3de76819cdf2c22625952d2dbc2e705b661e/cloud/gce/config.sh.tpl#L5)):*
Done.

---

*[cloud/gce/samples/3_node_cluster.tfvars, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEk3_KL1D0LSj4acNE9:-KEl6d1t0honOGEGKIsY:990158883) ([raw file](https://github.com/cockroachdb/cockroach/blob/b59c3de76819cdf2c22625952d2dbc2e705b661e/cloud/gce/samples/3_node_cluster.tfvars#L34)):*
Done. Yes, I recently switched to VSCode, so still ironing out some kinks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KEl9JcfNtMkwXYOdRwZ:1426412226)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Also, I agree that the AWS and GCE terraform configs should be functionally equivalent. However, I don't know if we've agreed upon the common requirements for both of them. For now, I think it's more fruitful to get more feedback and evolve the GCE config through actual use. I'm using it for my work on the rebalancing heuristics.

---

Review status: 2 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KElu6f1zJurHRn8-EQY:889449208)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
No, don't think we did. I'm fine with an iterative approach (just keeping in mind that the AWS one needs to tag along). Looking forward to playing with this myself (I want to get back to the acceptance test driven benchmark stuff a little more).

---

Review status: 2 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KEmMAVpDSTp7lsDZWZJ:1911681510)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890)*
<!-- Sent from Reviewable.io -->

mberhault commented on cloud/gce/launch.sh:
```diff
@@ -2,38 +2,33 @@
 # Setup iptables rules.
-sudo iptables -t nat -A PREROUTING -p tcp --dport ${PORT} -j DNAT --to-destination ${LOCAL_ADDRESS}
-sudo iptables -t nat -A POSTROUTING -p tcp -d ${LB_ADDRESS} --dport ${PORT} -j SNAT --to-source ${LOCAL_ADDRESS}
+sudo iptables -t nat -A PREROUTING -p tcp --dport ${SQL_PORT} -j DNAT --to-destination ${LOCAL_ADDRESS}
```
I would drop all the iptables stuff, it's completely broken anyway.

mberhault commented on cloud/gce/config.sh.tpl:
```diff
@@ -0,0 +1,5 @@
+SQL_PORT=${sql_port}
+HTTP_PORT=${http_port}
+LB_ADDRESS=${load_balancer_address}
```
you can drop all the load_balancer stuff, it's not needed anymore.

mberhault commented on cloud/gce/main.tf:
```diff
@@ -32,30 +32,77 @@ resource "google_compute_instance" "cockroach" {
     destination = "/home/ubuntu/launch.sh"
   }
 
+  provisioner "file" {
```
I would drop this and the cockroach_binary variable. We can already specify the sha, that should be sufficient, and probably still easier than building a linux binary for most people.

tschottdorf replied:
I think @cuongdo mentioned that that was the mode he wanted to use. I also think it could be useful to spin up a cluster from a local WIP without a hassle.

mberhault replied:
fair enough, but most people (even among us) don't build linux binaries.

tschottdorf replied:
That's not an argument because the only reason we don't do it is because we have no use for them. This creates a use and `./build/builder.sh make build` really isn't that hard.

cuongdo replied:
Yes, building Linux binaries on OS X using Docker is easy, and it's useful
for testing things on a "real" cluster such as rebalancing. I'm keeping the
option to use your own binary in, since the default behavior of downloading
the binary should work well for the typical user.
On Mon, Apr 11, 2016 at 8:16 AM Tobias Schottdorf <notifications@github.com>
wrote:

> In cloud/gce/main.tf
> <https://github.com/cockroachdb/cockroach/pull/5890#discussion_r59194062>:
>
> >    }
> >
> >    provisioner "file" {
> >      source = "launch.sh"
> >      destination = "/home/ubuntu/launch.sh"
> >    }
> >
> > +  provisioner "file" {
>
> That's not an argument because the only reason we don't do it is because
> we have no use for them. This creates a use and ./build/builder.sh make
> build really isn't that hard.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5890/files/391d9cde36f49fdde1375dc1a53155f723ca0ccb#r59194062>
>

mberhault commented on cloud/gce/samples/3_node_cluster.tfvars:
```diff
@@ -0,0 +1,35 @@
+# This configuration starts a 3-node CockroachDB cluster. You'll want to
```
I don't think you need this file. `num_instances` will pretty much always be specified on the command line, and the others still need to be filled in anyway.

cuongdo commented:
Yes, definitely (re: AWS config keeping pace).

---

Review status: 4 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[cloud/gce/config.sh.tpl, line 3 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9byr-r2-3:-KF4vFgmwb3PBOZf-_9K:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/391d9cde36f49fdde1375dc1a53155f723ca0ccb/cloud/gce/config.sh.tpl#L3)):*
Done.

---

*[cloud/gce/launch.sh, line 23 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9bym-r2-23:-KF4vJZppU7XqV8Xo3Nc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/391d9cde36f49fdde1375dc1a53155f723ca0ccb/cloud/gce/launch.sh#L23)):*
Done.

---

*[cloud/gce/samples/3_node_cluster.tfvars, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9bys-r2-1:-KF4v_0WaO8yIpyBDnnc:577247598) ([raw file](https://github.com/cockroachdb/cockroach/blob/391d9cde36f49fdde1375dc1a53155f723ca0ccb/cloud/gce/samples/3_node_cluster.tfvars#L1)):*
Fair enough, I've removed it and updated the documentation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KF50pbgJphHeRJOVjw0:-1918641566)*
<!-- Sent from Reviewable.io -->

mberhault commented on cloud/gce/launch.sh:
```diff
@@ -2,38 +2,27 @@
 action=$1
-if [ "${action}" != "init" -a "${action}" != "start" ]; then
-  echo "Usage: ${0} [init|start] [load balancer address:port]"
+if [ "${action}" != "start" ]; then
```
since we only have one possible action now, this can be removed, along with the `action` variable, the call in `main.tf` and the readme.

mberhault commented on cloud/gce/README.md:
```diff
@@ -1,3 +1,109 @@
+  admin_urls   = http://127.196.142.127:8080/, http://127.196.17.15:8080/, http://127.196.139.40:8080/
+  instances    = cockroach-0, cockroach-1, cockroach-2
+  sql_lb_url   = postgresql://root@127.196.119.186:26257/?sslmode=disable
+  sql_urls     = postgresql://root@$127.196.142.127:26257/?sslmode=disable, postgresql://root@$127.196.17.15:26257/?sslmode=disable, postgresql://root@$127.196.139.40:26257/?sslmode=disable
```
this looks odd. what's with the `$` at the beginning of the ip address?

mberhault commented on cloud/gce/output.tf:
```diff
@@ -1,15 +1,25 @@
 
+# Postgres URLs for individual instances.
+output "sql_urls" {
+  value = "${join(", ", formatlist("postgresql://root@$%s:%s/?sslmode=disable", google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip, var.sql_port))}"
```
ah. right. this is where it comes from. no `$` in the fmt spec.

mberhault commented on cloud/gce/network.tf:
```diff
@@ -19,7 +19,9 @@ resource "google_compute_target_pool" "default" {
 resource "google_compute_forwarding_rule" "default" {
```
I think almost all the `network.tf` stuff can go away, though we may still need the firewall rule.

cuongdo commented:
Review status: 2 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[cloud/gce/launch.sh, line 23 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9bym-r2-23:-KF5qa-fkZ4jheFR1ipi:-16608034) ([raw file](https://github.com/cockroachdb/cockroach/blob/391d9cde36f49fdde1375dc1a53155f723ca0ccb/cloud/gce/launch.sh#L23)):*
Actually, I had to keep the PREROUTING rules to allow the `google_compute_forwarding_rule` to work.

---

*[cloud/gce/launch.sh, line 17 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9bym-r3-17:-KF5qFmahL_9KNaKZji3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a54ae41c865a01c2a124e1406369ebfa440eb5ea/cloud/gce/launch.sh#L17)):*
Done.

---

*[cloud/gce/network.tf, line 19 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9byo-r3-19:-KF5TZp2yfbUJVVcr7ts:2114106261) ([raw file](https://github.com/cockroachdb/cockroach/blob/a54ae41c865a01c2a124e1406369ebfa440eb5ea/cloud/gce/network.tf#L19)):*
What specifically can go away?

I like having `google_compute_forwarding_rule`, because that provides me with a single load-balanced URL to access. Also, most "real" installations will likely some kind of load balancing in front of CockroachDB.

The forwarding rule also makes load generation across the cluster much easier, since you can point the load generation tool at the forwarding IP, and the client traffic gets spread out.

---

*[cloud/gce/output.tf, line 18 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9byp-r3-18:-KF5UA-46q94ZvK3MfiG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a54ae41c865a01c2a124e1406369ebfa440eb5ea/cloud/gce/output.tf#L18)):*
Done.

---

*[cloud/gce/README.md, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9byk-r3-51:-KF5U8GvA3PgyQsxsSiF:1758359711) ([raw file](https://github.com/cockroachdb/cockroach/blob/a54ae41c865a01c2a124e1406369ebfa440eb5ea/cloud/gce/README.md#L51)):*
Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 2 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[cloud/gce/network.tf, line 19 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-KEgENzbdJazRnos9byo-r3-19:-KF5rHqWMFiFmC2PzFbQ:-1037087142) ([raw file](https://github.com/cockroachdb/cockroach/blob/a54ae41c865a01c2a124e1406369ebfa440eb5ea/cloud/gce/network.tf#L19)):*
that's fine, but that only works for http. The TCP load balancer in GCE is completely wonky, and will cause problems for the postgres protocol anyway.
So all the iptables stuff can go away, and the fowarding rule should be limited to the http port.
At that point, the load balancer is only useful to reach an arbitrary node's UI, and the name isn't much better than using any one of the nodes. But do feel free to leave it in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890)*
<!-- Sent from Reviewable.io -->

mberhault commented:
LGTM. I presume you've checked, and it still works :)

cuongdo commented:
What problems have you seen with the GCE TCP load balancer?

On Mon, Apr 11, 2016 at 3:51 PM marc <notifications@github.com> wrote:

> LGTM. I presume you've checked, and it still works :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5890#issuecomment-208527531>
>

mberhault commented:
the main one is that is fakes the ELB address.
Specifically, node1, node2, etc... will have routing rules that makes them see elb_address as theirs. This is fine to accept incoming connections, but it means that the nodes can't use the ELB address to reach other nodes. That's what the iptables crap was, I was trying to get around that.

On the postgres side, the danger is in jumping from one node to another. It would need to be a new connection, and postgres doesn't handle that.

cuongdo commented:
I've removed the `sql_urls` output variable, which uses the forwarding rule to forward SQL requests.

---

Review status: 2 of 9 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5890#-:-KFBKVGJAImE67v0ShM2:-1236906494)*
<!-- Sent from Reviewable.io -->
